File: packages/dvai-bridge-flutter/android/src/main/kotlin/co/deepvoiceai/bridge/flutter/DVAIBridgeFlutterPlugin.kt
package co.deepvoiceai.bridge.flutter

// DVAIBridgeFlutterPlugin — Kotlin entry point for the `dvai_bridge`
// Flutter plugin (Android side).
//
// Architecture mirrors the iOS plugin:
//
//   - Implements the Pigeon-generated `DVAIBridgeHostApi` interface (4
//     lifecycle methods). Each method launches a coroutine on
//     `pluginScope` to call into `co.deepvoiceai.bridge.DVAIBridge`'s
//     suspend functions, then bridges the result back into Pigeon's
//     `Result<…>` callback.
//
//   - Subscribes to `DVAIBridge.progressFlow` on first listener of the
//     Pigeon event channel and emits `ProgressEventMessage` payloads.
//     Cancels the collection job on `onCancel`.
//
//   - Translates Kotlin's sealed `DVAIBridgeError` hierarchy into
//     `FlutterError` instances whose `code` field uses the same lowercase
//     camelCase wire identifier as the Dart enum. The Dart facade decodes
//     back via `DVAIBridgeError.fromPlatform(...)`.
//
// Lifecycle: one plugin instance per `FlutterEngine`. `onAttachedToEngine`
// wires the Pigeon channels; `onDetachedFromEngine` tears them down and
// cancels the plugin scope.

import co.deepvoiceai.bridge.BackendKind
import co.deepvoiceai.bridge.BoundServer
import co.deepvoiceai.bridge.DVAIBridge
import co.deepvoiceai.bridge.DVAIBridgeError
import co.deepvoiceai.bridge.DownloadOptions
import co.deepvoiceai.bridge.ProgressEvent
import co.deepvoiceai.bridge.StartOptions
import co.deepvoiceai.bridge.StatusInfo
import co.deepvoiceai.bridge.shared.core.CorsConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class DVAIBridgeFlutterPlugin : FlutterPlugin, DVAIBridgeHostApi {

    private var pluginScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var binaryMessenger: BinaryMessenger? = null
    private val progressStreamHandler = ProgressStreamHandler()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // Re-init defensively. Idempotent in DVAIBridge.init.
        DVAIBridge.init(binding.applicationContext)

        val messenger = binding.binaryMessenger
        binaryMessenger = messenger

        DVAIBridgeHostApi.setUp(messenger, this)
        ProgressEventsStreamHandler.register(messenger, progressStreamHandler)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        binaryMessenger?.let { messenger ->
            DVAIBridgeHostApi.setUp(messenger, null)
        }
        binaryMessenger = null
        progressStreamHandler.stopCollection()
        pluginScope.cancel()
        // Recreate so a re-attach (rare; e.g. add-to-app `FlutterEngineGroup`
        // scenarios) gets a fresh scope.
        pluginScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    }

    // MARK: - DVAIBridgeHostApi

    override fun startBridge(
        opts: StartOptionsMessage,
        callback: (Result<BoundServerMessage>) -> Unit,
    ) {
        pluginScope.launch {
            try {
                val startOptions = toStartOptions(opts)
                val server = DVAIBridge.start(startOptions)
                callback(Result.success(toMessage(server)))
            } catch (e: DVAIBridgeError) {
                callback(Result.failure(toFlutterError(e)))
            } catch (e: Throwable) {
                callback(
                    Result.failure(
                        toFlutterError(DVAIBridgeError.BackendError(e)),
                    ),
                )
            }
        }
    }

    override fun stopBridge(callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                DVAIBridge.stop()
                callback(Result.success(Unit))
            } catch (e: DVAIBridgeError) {
                callback(Result.failure(toFlutterError(e)))
            } catch (e: Throwable) {
                callback(
                    Result.failure(
                        toFlutterError(DVAIBridgeError.BackendError(e)),
                    ),
                )
            }
        }
    }

    override fun status(callback: (Result<StatusInfoMessage>) -> Unit) {
        // status() is synchronous on Android — wrap for parity with iOS.
        try {
            val info: StatusInfo = DVAIBridge.status()
            callback(
                Result.success(
                    StatusInfoMessage(
                        running = info.running,
                        baseUrl = info.baseUrl,
                        port = null,
                        backend = info.backend?.toWire(),
                        modelId = info.modelId,
                    ),
                ),
            )
        } catch (e: Throwable) {
            callback(
                Result.failure(
                    toFlutterError(DVAIBridgeError.BackendError(e)),
                ),
            )
        }
    }

    override fun downloadModel(
        opts: DownloadOptionsMessage,
        callback: (Result<DownloadResultMessage>) -> Unit,
    ) {
        pluginScope.launch {
            try {
                val destFilename = opts.destFilename ?: deriveFilename(opts.url)
                val result = DVAIBridge.downloadModel(
                    DownloadOptions(
                        url = opts.url,
                        sha256 = opts.sha256,
                        destFilename = destFilename,
                    ),
                )
                callback(
                    Result.success(
                        DownloadResultMessage(
                            path = result.path,
                            sha256 = result.sha256,
                            sizeBytes = result.sizeBytes,
                            cached = null,
                        ),
                    ),
                )
            } catch (e: DVAIBridgeError) {
                callback(Result.failure(toFlutterError(e)))
            } catch (e: Throwable) {
                callback(
                    Result.failure(
                        toFlutterError(
                            DVAIBridgeError.DownloadFailed(
                                e.message ?: "unknown",
                                e,
                            ),
                        ),
                    ),
                )
            }
        }
    }

    private fun deriveFilename(url: String): String {
        val trimmed = url.substringBefore('?').substringBefore('#')
        return trimmed.substringAfterLast('/').ifEmpty { "model.bin" }
    }

    // MARK: - Translation helpers

    private fun toStartOptions(msg: StartOptionsMessage): StartOptions {
        val backend = backendKindFromWire(msg.backend)
            ?: throw DVAIBridgeError.ConfigurationInvalid("Unknown backend ${msg.backend}")

        val cors: CorsConfig = when (val raw = msg.corsOrigin) {
            null, "*" -> CorsConfig.Wildcard
            else -> if (raw.contains(",")) {
                CorsConfig.Allowlist(raw.split(",").map { it.trim() }.filter { it.isNotEmpty() })
            } else {
                CorsConfig.Exact(raw)
            }
        }

        return StartOptions(
            backend = backend,
            modelPath = msg.modelPath,
            tokenizerPath = msg.tokenizerPath,
            mmprojPath = msg.mmprojPath,
            chatTemplate = msg.chatTemplate,
            gpuLayers = msg.gpuLayers?.toInt() ?: 99,
            contextSize = msg.contextSize?.toInt() ?: 2048,
            threads = msg.threads?.toInt() ?: 4,
            embeddingMode = msg.embeddingMode ?: false,
            visionEnabled = msg.visionEnabled ?: false,
            temperature = msg.temperature?.toFloat() ?: 0f,
            topP = msg.topP?.toFloat() ?: 1f,
            topK = msg.topK?.toInt() ?: 0,
            maxNewTokens = msg.maxNewTokens?.toInt() ?: 512,
            httpBasePort = msg.httpBasePort?.toInt() ?: 38883,
            httpMaxPortAttempts = msg.httpMaxPortAttempts?.toInt() ?: 16,
            corsOrigin = cors,
            modelId = msg.modelId,
        )
    }

    private fun toMessage(server: BoundServer): BoundServerMessage {
        return BoundServerMessage(
            baseUrl = server.baseUrl,
            port = server.port.toLong(),
            backend = server.backend.toWire(),
            modelId = server.modelId,
        )
    }

    private fun toFlutterError(error: DVAIBridgeError): FlutterError {
        val code: String
        val details: Map<String, Any?>?
        when (error) {
            is DVAIBridgeError.AlreadyStarted -> {
                code = "alreadyStarted"
                details = mapOf(
                    "backend" to error.currentBackend.toWire(),
                    "baseUrl" to error.baseUrl,
                )
            }
            is DVAIBridgeError.ConfigurationInvalid -> {
                code = "configurationInvalid"
                details = null
            }
            is DVAIBridgeError.ModelLoadFailed -> {
                code = "modelLoadFailed"
                details = null
            }
            is DVAIBridgeError.BackendUnavailable -> {
                code = "backendUnavailable"
                details = mapOf("backend" to error.backend.toWire())
            }
            is DVAIBridgeError.BackendError -> {
                code = "backendError"
                details = null
            }
            is DVAIBridgeError.ChecksumMismatch -> {
                code = "checksumMismatch"
                // The class doesn't expose `expected` / `actual` as
                // properties post-construction; the message string carries
                // them. Pass through as a single message instead.
                details = null
            }
            is DVAIBridgeError.DownloadFailed -> {
                code = "downloadFailed"
                details = null
            }
        }
        return FlutterError(
            code = code,
            message = error.message ?: error.toString(),
            details = details,
        )
    }

    // MARK: - Progress event channel

    /// Stream handler for the Pigeon `progressEvents` event channel. Collects
    /// `DVAIBridge.progressFlow` on plugin scope and emits the
    /// `ProgressEventMessage` shape that the iOS side also produces.
    inner class ProgressStreamHandler : ProgressEventsStreamHandler() {
        private var collectionJob: Job? = null

        override fun onListen(p0: Any?, sink: PigeonEventSink<ProgressEventMessage>) {
            collectionJob?.cancel()
            collectionJob = pluginScope.launch(Dispatchers.Main.immediate) {
                DVAIBridge.progressFlow.collect { event ->
                    sink.success(toMessage(event))
                }
            }
        }

        override fun onCancel(p0: Any?) {
            collectionJob?.cancel()
            collectionJob = null
        }

        fun stopCollection() {
            collectionJob?.cancel()
            collectionJob = null
        }

        private fun toMessage(event: ProgressEvent): ProgressEventMessage {
            return when (event) {
                is ProgressEvent.Started -> ProgressEventMessage(
                    kind = "started",
                    phase = event.phase,
                )
                is ProgressEvent.Progress -> ProgressEventMessage(
                    kind = "progress",
                    phase = event.phase,
                    percent = event.percent.toDouble().takeIf { it >= 0.0 }
                        ?.let { it * 100.0 },
                    message = event.message.ifEmpty { null },
                )
                is ProgressEvent.Completed -> ProgressEventMessage(
                    kind = "completed",
                    phase = event.phase,
                )
                is ProgressEvent.Failed -> ProgressEventMessage(
                    kind = "failed",
                    phase = event.phase,
                    errorKind = event.error.toWireKind(),
                    errorMessage = event.error.message,
                )
            }
        }
    }
}

// MARK: - Wire-format helpers

private fun BackendKind.toWire(): String = when (this) {
    BackendKind.Auto -> "auto"
    BackendKind.Llama -> "llama"
    BackendKind.MediaPipe -> "mediapipe"
    BackendKind.LiteRT -> "litert"
}

private fun backendKindFromWire(value: String): BackendKind? = when (value) {
    "auto" -> BackendKind.Auto
    "llama" -> BackendKind.Llama
    "mediapipe" -> BackendKind.MediaPipe
    "litert" -> BackendKind.LiteRT
    else -> null
}

private fun DVAIBridgeError.toWireKind(): String = when (this) {
    is DVAIBridgeError.AlreadyStarted -> "alreadyStarted"
    is DVAIBridgeError.ConfigurationInvalid -> "configurationInvalid"
    is DVAIBridgeError.ModelLoadFailed -> "modelLoadFailed"
    is DVAIBridgeError.BackendUnavailable -> "backendUnavailable"
    is DVAIBridgeError.BackendError -> "backendError"
    is DVAIBridgeError.ChecksumMismatch -> "checksumMismatch"
    is DVAIBridgeError.DownloadFailed -> "downloadFailed"
}

File: packages/dvai-bridge-flutter/ios/Classes/DVAIBridgeFlutterPlugin.swift
// DVAIBridgeFlutterPlugin — Swift entry point for the `dvai_bridge`
// Flutter plugin.
//
// Architecture:
//
//   - Implements the Pigeon-generated `DVAIBridgeHostApi` protocol (4 lifecycle
//     methods). Each method opens a `Task { ... }`, awaits the underlying
//     `DVAIBridge.shared` actor, and bridges the result back to Pigeon's
//     completion handler. This is the documented pattern from the Phase 3F
//     spec §6 question 9 — Pigeon doesn't generate `actor` methods directly
//     (as of 26.3.4), so we wrap.
//
//   - Subscribes to `DVAIBridge.shared.progressPublisher` (Combine) on first
//     listener and pumps events through the Pigeon-generated event channel
//     (`ProgressEventsStreamHandler`).
//
//   - Translates Swift's `DVAIBridgeError` cases into Pigeon `PigeonError`
//     instances whose `code` field uses the same lowercase camelCase wire
//     identifier as the Dart enum (`DVAIBridgeErrorKind.wireValue`). The
//     Dart facade decodes back via `DVAIBridgeError.fromPlatform(...)`.
//
// The plugin instance lifetime is tied to the Flutter engine. Multiple
// engines (e.g. add-to-app scenarios with `FlutterEngineGroup`) each get
// their own `DVAIBridgeFlutterPlugin` instance and their own progress
// subscription, but they share the singleton `DVAIBridge.shared` actor —
// the Phase 3C SDK's design.

import Combine
import Flutter
import Foundation
import UIKit

import DVAIBridge

public final class DVAIBridgeFlutterPlugin: NSObject, FlutterPlugin, DVAIBridgeHostApi {
  /// Combine subscription to `DVAIBridge.shared.progressPublisher`. Held as
  /// a property so the plugin instance retains it for its lifetime; the
  /// stream handler installs/replaces it on `onListen`.
  private var progressCancellable: AnyCancellable?

  /// Strong reference to the Pigeon stream handler so its lifetime matches
  /// the plugin instance (Flutter doesn't retain stream handlers itself).
  private var progressStreamHandler: ProgressStreamHandler?

  public static func register(with registrar: any FlutterPluginRegistrar) {
    let plugin = DVAIBridgeFlutterPlugin()
    DVAIBridgeHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: plugin)

    let streamHandler = ProgressStreamHandler()
    plugin.progressStreamHandler = streamHandler
    ProgressEventsStreamHandler.register(
      with: registrar.messenger(),
      streamHandler: streamHandler
    )
  }

  // MARK: - DVAIBridgeHostApi

  public func startBridge(
    opts: StartOptionsMessage,
    completion: @escaping (Result<BoundServerMessage, Error>) -> Void
  ) {
    Task {
      do {
        let config = try DVAIBridgeFlutterPlugin.makeConfig(from: opts)
        let server = try await DVAIBridge.shared.start(config)
        completion(.success(DVAIBridgeFlutterPlugin.toMessage(server)))
      } catch let bridgeError as DVAIBridgeError {
        completion(.failure(DVAIBridgeFlutterPlugin.toPigeonError(bridgeError)))
      } catch {
        completion(.failure(PigeonError(
          code: "backendError",
          message: error.localizedDescription,
          details: nil
        )))
      }
    }
  }

  public func stopBridge(completion: @escaping (Result<Void, Error>) -> Void) {
    Task {
      do {
        try await DVAIBridge.shared.stop()
        completion(.success(()))
      } catch let bridgeError as DVAIBridgeError {
        completion(.failure(DVAIBridgeFlutterPlugin.toPigeonError(bridgeError)))
      } catch {
        completion(.failure(PigeonError(
          code: "backendError",
          message: error.localizedDescription,
          details: nil
        )))
      }
    }
  }

  public func status(
    completion: @escaping (Result<StatusInfoMessage, Error>) -> Void
  ) {
    Task {
      let info = await DVAIBridge.shared.status()
      completion(.success(StatusInfoMessage(
        running: info.running,
        baseUrl: info.baseUrl,
        port: nil,
        backend: info.backend?.rawValue,
        modelId: nil
      )))
    }
  }

  public func downloadModel(
    opts: DownloadOptionsMessage,
    completion: @escaping (Result<DownloadResultMessage, Error>) -> Void
  ) {
    Task {
      do {
        guard let url = URL(string: opts.url) else {
          throw DVAIBridgeError.configurationInvalid(reason: "Invalid URL: \(opts.url)")
        }
        var headers: [String: String] = [:]
        if let pairs = opts.headers {
          var index = 0
          while index + 1 < pairs.count {
            if let key = pairs[index], let value = pairs[index + 1] {
              headers[key] = value
            }
            index += 2
          }
        }
        let downloadOpts = DVAIBridge.DownloadOptions(
          url: url,
          sha256: opts.sha256,
          destFilename: opts.destFilename,
          headers: headers
        )
        let result = try await DVAIBridge.shared.downloadModel(downloadOpts)
        // The iOS DVAIBridge.DownloadResult exposes `path` + `cached` only;
        // sha256 / sizeBytes aren't surfaced, so echo the request's sha256
        // and 0 bytes so the cross-platform shape stays consistent. The
        // Android side fills in the real numbers; iOS-side smoke tests for
        // those fields can read them off-band via Foundation if needed.
        completion(.success(DownloadResultMessage(
          path: result.path,
          sha256: opts.sha256,
          sizeBytes: 0,
          cached: result.cached
        )))
      } catch let bridgeError as DVAIBridgeError {
        completion(.failure(DVAIBridgeFlutterPlugin.toPigeonError(bridgeError)))
      } catch {
        completion(.failure(PigeonError(
          code: "downloadFailed",
          message: error.localizedDescription,
          details: nil
        )))
      }
    }
  }

  public func assessHardware(
    hardwareMinimum: Double,
    minLocalCapability: Double,
    completion: @escaping (Result<HardwareAssessmentMessage, Error>) -> Void
  ) {
    let a = DVAIBridge.shared.assessHardware(
      hardwareMinimum: hardwareMinimum,
      minLocalCapability: minLocalCapability
    )
    completion(.success(HardwareAssessmentMessage(
      mode: a.mode.rawValue,
      tokPerSec: a.tokPerSec,
      reason: a.reason,
      hasNpu: a.hints.hasNpu,
      ramGb: Int64(a.hints.ramGb),
      gpuClass: a.hints.gpuClass.rawValue,
      cpuClass: a.hints.cpuClass.rawValue
    )))
  }

  // MARK: - Translation helpers

  private static func makeConfig(from msg: StartOptionsMessage) throws -> DVAIBridgeConfig {
    guard let backend = BackendKind(rawValue: msg.backend) else {
      throw DVAIBridgeError.configurationInvalid(reason: "Unknown backend \(msg.backend)")
    }
    let cors: DVAIBridgeConfig.CORSOrigin
    if let raw = msg.corsOrigin {
      if raw == "*" {
        cors = .wildcard
      } else if raw.contains(",") {
        let allowlist = raw
          .split(separator: ",")
          .map { String($0).trimmingCharacters(in: .whitespaces) }
          .filter { !$0.isEmpty }
        cors = .allowlist(allowlist)
      } else {
        cors = .exact(raw)
      }
    } else {
      cors = .wildcard
    }
    return DVAIBridgeConfig(
      backend: backend,
      modelPath: msg.modelPath,
      mmprojPath: msg.mmprojPath,
      tokenizerPath: msg.tokenizerPath,
      gpuLayers: msg.gpuLayers.map(Int.init) ?? 99,
      contextSize: msg.contextSize.map(Int.init) ?? 2048,
      threads: msg.threads.map(Int.init) ?? 4,
      embeddingMode: msg.embeddingMode ?? false,
      httpBasePort: msg.httpBasePort.map(Int.init) ?? 38883,
      httpMaxPortAttempts: msg.httpMaxPortAttempts.map(Int.init) ?? 16,
      corsOrigin: cors,
      autoUnloadOnLowMemory: msg.autoUnloadOnLowMemory ?? false,
      logLevel: msg.logLevel ?? "info"
    )
  }

  private static func toMessage(_ server: BoundServer) -> BoundServerMessage {
    return BoundServerMessage(
      baseUrl: server.baseUrl,
      port: Int64(server.port),
      backend: server.backend.rawValue,
      modelId: server.modelId
    )
  }

  private static func toPigeonError(_ error: DVAIBridgeError) -> PigeonError {
    let code: String
    var details: [String: Any]? = nil
    switch error {
    case .alreadyStarted(let backend, let baseUrl):
      code = "alreadyStarted"
      details = ["backend": backend.rawValue, "baseUrl": baseUrl]
    case .notStarted:
      code = "notStarted"
    case .configurationInvalid:
      code = "configurationInvalid"
    case .backendUnavailable(let backend, _):
      code = "backendUnavailable"
      details = ["backend": backend.rawValue]
    case .modelLoadFailed:
      code = "modelLoadFailed"
    case .downloadFailed:
      code = "downloadFailed"
    case .checksumMismatch:
      code = "checksumMismatch"
    case .backendError:
      code = "backendError"
    }
    return PigeonError(
      code: code,
      message: error.errorDescription,
      details: details
    )
  }

  // MARK: - Progress event channel

  /// Stream handler for the Pigeon `progressEvents` event channel. Subscribes
  /// to `DVAIBridge.shared.progressPublisher` on first listener and replays
  /// events as `ProgressEventMessage` payloads. Cancels on `onCancel`.
  final class ProgressStreamHandler: PigeonEventChannelWrapper<ProgressEventMessage> {
    private var cancellable: AnyCancellable?

    override func onListen(
      withArguments arguments: Any?,
      sink: PigeonEventSink<ProgressEventMessage>
    ) {
      cancellable?.cancel()
      cancellable = DVAIBridge.shared.progressPublisher.sink { event in
        sink.success(ProgressStreamHandler.toMessage(event))
      }
    }

    override func onCancel(withArguments arguments: Any?) {
      cancellable?.cancel()
      cancellable = nil
    }

    /// Translate the iOS-side `ProgressEvent` into the Pigeon wire shape.
    /// The iOS SDK currently emits `Phase` only (no `kind`), so we infer
    /// `kind` from the phase: `error` → failed, `ready` → completed
    /// (post-start), `verify` → completed (post-download), and the rest →
    /// progress. The Android side already emits explicit kinds, so the
    /// Dart facade gets a uniform shape across both platforms.
    static func toMessage(_ event: ProgressEvent) -> ProgressEventMessage {
      let phase = event.phase.rawValue
      let kind: String
      switch event.phase {
      case .ready:
        kind = "completed"
      case .error:
        kind = "failed"
      case .verify:
        kind = "completed"
      case .download, .load:
        kind = "progress"
      }
      return ProgressEventMessage(
        kind: kind,
        phase: phase,
        percent: event.percent,
        message: event.message,
        errorKind: event.phase == .error ? "backendError" : nil,
        errorMessage: event.phase == .error ? event.message : nil
      )
    }
  }
}
